Swift: detect the use of constant passwords for password-based encryption

karimhamdanali opened the pull request:
Deriving password-based encryption keys using hard-coded passwords is insecure, because the generated key may be easily discovered. Data hashed using constant salts are vulnerable to dictionary attacks, enabling attackers to recover the original input.

The rule currently supports all ciphers that the CryptoSwift API provides, but we can always extend it further if more APIs are added.

I'd appreciate a review of the query itself, the accompanying tests, and the associated documentation.

geoffw0 commented on swift/ql/src/queries/Security/CWE-259/ConstantPassword.ql:
```diff
@@ -0,0 +1,64 @@
+import DataFlow::PathGraph
+
+/**
+ * A constant salt is created through either a byte array or string literals.
```
```suggestion
 * A constant password is created through either a byte array or string literals.
```

karimhamdanali replied:
Ah good catch! I thought I initially got them all.

geoffw0 commented on swift/ql/src/queries/Security/CWE-259/ConstantPassword.ql:
```diff
@@ -0,0 +1,64 @@
+
+/**
+ * A taint configuration from the source of constants passwords to expressions that use
+ * them to initialize password-based enecryption keys.
```
```suggestion
 * them to initialize password-based encryption keys.
```

karimhamdanali replied:
fixed as suggested

geoffw0 reviewed (commented):
The query looks good, though like https://github.com/github/codeql/pull/10993 I believe we should have an issue to expand coverage to other libraries / APIs in future.  The documentation is currently a little bit confusing, I think just because bits are left over from the other query.

karimhamdanali commented on swift/ql/src/queries/Security/CWE-259/ConstantPassword.qhelp:
```diff
@@ -0,0 +1,21 @@
+  </example>
+
+  <references>
+    <li><a href="https://www.okta.com/blog/2019/03/what-are-salted-passwords-and-password-hashing/">What are Salted Passwords and Password Hashing?</a></li>
```
[This RFC](https://www.rfc-editor.org/rfc/rfc2898) is more about the process where these passwords are stored. Do you think it might be more appropriate here?

geoffw0 replied:
That looks like a good reference.  No harm in having both if both are relevant.

karimhamdanali replied:
I will keep both then.

geoffw0 commented on swift/ql/src/queries/Security/CWE-259/ConstantPassword.qhelp:
```diff
@@ -0,0 +1,21 @@
+  "qhelp.dtd">
+<qhelp>
+  <overview>
+    <p>Deriving password-based encryption keys using hard-coded passwords is insecure, because the generated key may be easily discovered. Data hashed using constant salts are vulnerable to dictionary attacks, enabling attackers to recover the original input.</p>
```
In that case perhaps the query `@description` needs updating, as that's where the reference "potential attackers can easily recover them from the source code" comes from.

On hardcoded credentials, we have queries like `cs/hardcoded-credentials` for this but also rely on secret scanning (which is separate from codeql) for textually recognizable things like AWS keys (regardless of whether they're in source code or not).

geoffw0 replied:
I'm still not sure I understand what is being said about salts and dictionary attacks - doesn't a constant password for key derivation enable more straightforward attacks than dictionary attacks?

karimhamdanali replied:
I'll take a look at the proposed `hardcoded-credentials` later today then.

karimhamdanali replied:
Constant passwords would enable easier recovery of the key, but there's still the salt that's used for the derivation process. If that salt is random enough, then recovery won't be as easy as just looking up a hardcoded credential in the source code.

geoffw0 replied:
Ah, that makes sense.  We want to be as clear as possible about this in the `.qhelp` and query `@description`.

karimhamdanali replied:
I've added further clarification to `.qhelp` so as not to cloud the @descrioption too much. Do you think this PR is now ready for docs review?

karimhamdanali replied:
Awesome!

mchammer01 commented on swift/ql/src/queries/Security/CWE-259/ConstantPassword.qhelp:
```diff
@@ -0,0 +1,23 @@
+
+  <references>
+    <li><a href="https://www.okta.com/blog/2019/03/what-are-salted-passwords-and-password-hashing/">What are Salted Passwords and Password Hashing?</a></li>
+    <li><a href="https://www.rfc-editor.org/rfc/rfc2898">Password-Based Cryptography Specification</a></li>
```
Missing full stop
```suggestion
    <li><a href="https://www.rfc-editor.org/rfc/rfc2898">Password-Based Cryptography Specification</a>.</li>
```

karimhamdanali replied:
fixed as suggested

mchammer01 commented on swift/ql/src/queries/Security/CWE-259/ConstantPassword.qhelp:
```diff
@@ -0,0 +1,23 @@
+  </example>
+
+  <references>
+    <li><a href="https://www.okta.com/blog/2019/03/what-are-salted-passwords-and-password-hashing/">What are Salted Passwords and Password Hashing?</a></li>
```
For these two top references, shall we mention the source, for example
(this is just a question)

`<li>Okta blog: <a href="https://www.okta.com/blog/2019/03/what-are-salted-passwords-and-password-hashing/">What are Salted Passwords and Password Hashing?</a></li>

karimhamdanali replied:
Agreed and I've changed it accordingly.

mchammer01 commented on swift/ql/src/queries/Security/CWE-259/ConstantPassword.qhelp:
```diff
@@ -0,0 +1,23 @@
+  "qhelp.dtd">
+<qhelp>
+  <overview>
+    <p>Deriving password-based encryption keys using hard-coded passwords is insecure, because the generated key may be easily discovered. Data hashed using constant salts are vulnerable to dictionary attacks, enabling attackers to recover the original input.</p>
```
I think we use `hardcoded` (one word) in our query help, but I could be wrong 🤔 
Also you use one word only on line 7, so probably best to be consistent.
```suggestion
    <p>Deriving password-based encryption keys using hardcoded passwords is insecure, because the generated key may be easily discovered. Data hashed using constant salts are vulnerable to dictionary attacks, enabling attackers to recover the original input.</p>
```

karimhamdanali replied:
Agreed. Fixed as suggested.

mchammer01 commented on swift/ql/src/queries/Security/CWE-259/ConstantPassword.qhelp:
```diff
@@ -0,0 +1,23 @@
+  "qhelp.dtd">
+<qhelp>
+  <overview>
+    <p>Deriving password-based encryption keys using hard-coded passwords is insecure, because the generated key may be easily discovered. Data hashed using constant salts are vulnerable to dictionary attacks, enabling attackers to recover the original input.</p>
```
I think the word data is generally treated as singular, but ok for you to leave as-is if you have strong feelings about this 😄 
 
```suggestion
    <p>Deriving password-based encryption keys using hard-coded passwords is insecure, because the generated key may be easily discovered. Data hashed using constant salts is vulnerable to dictionary attacks, enabling attackers to recover the original input.</p>
```

karimhamdanali replied:
I don't :D Changed as suggested.

mchammer01 reviewed (dismissed):
@karimhamdanali - this LGTM ✨ 
I've added a few comments and suggestions. Feel free to ignore them if you don't agree 😅

karimhamdanali commented:
@mchammer01 Thanks again for the quick docs review. I think I've addressed all comments now. Please let me know if there's anything else I need to address before being able to merge this PR :)

mchammer01 reviewed (approved):
LGTM, thanks for addressing my comments 💖
